std/string.cup

```diff
@@ -0,0 +1,268 @@
+method str::startswith_cstr(prefix: char*): i32{
+    return strstartswith(self.chars, prefix);
+}
+
+method str::startswith(prefix: str*): i32{
+    return self::startswith_cstr(prefix.chars);
+}
+
+method str::equals_cstr(other: char*): i32{
+    return streq(self.chars, other);
+}
+
+method str::equals(other: str*): i32{
+    return self::equals_cstr(other.chars);
+}
+
+// Checks whether self contains other and returns index of start of the other in self if it contains other.
+method str::contains_idx(other: str*): i32{
+    if (other.len == 0){
+        die(here, "str::contains: other has to have length greater than 0");
+    }
+    
+    for (let i = 0; i < self.len; ++i){
+        if (self::get(i) == other::get(0)){
+            let to = i+other.len-1;
+            if (to >= self.len){
+                break;
+            }
+            let slice_: slice* = self::slice(i, to);
+            
+            let contains = true;
+            for (let i_ = 0; i_ < slice_.len; ++i_){
+                if (slice_::get(i_) != other::get(i_)){
+                    contains = false;
+                    break;
+                }
+            }
+            
+            if (contains){
+                return i;
+            }
+            
+            free(slice_);
+        }
+    }
+    
+    return -1;
+}
+
+method str::contains(other: str*): i32{
+    let result = self::contains_idx(other);
+    
+    if (result == -1){
+        return false;
+    }
+    else 
+        return true;
+}
+
+method str::drop();
+
+method str::contains_cstr(other: char*): i32{
+    let temp_str: str* = new str(other);
+    let result = self::contains(temp_str);
+    temp_str::drop();
+    free(temp_str);
+    if (result == -1){
+        return false;
+    }
+    else 
+        return true;
+}
+
+// Removes first occurance of other and creates a new string from resulting string 
+method str::remove(other: str*): str*{
+    let idx = self::contains_idx(other);
+    
+    if (idx != -1){
+        if (idx-1 < 0){
+            if (idx+other.len < self.len){
+                let second = self::slice(idx+other.len, self.len);
+                let new_str = second::to_str();
+                return new_str;
+            } else{
+                return new str("");
+            }
+        }
+        else{
+            let first = self::slice(0, idx-1);
+            if (idx+other.len < self.len){
+                let second = self::slice(idx+other.len, self.len);
+                let new_str = first::to_str();
+                new_str::push_slice(second);
+                free(second);
+                free(first);
+                return new_str;
+            }
+            else{
+                return first::to_str();
+            }
+        }
+    }
+    
+    return self;
+}
+
+// Implicit destructor
+// You should call this before freeing str itself, if you don't do this you will get a memory leak.
+method str::drop(){
+    free(self.chars);
+    self.isdropped = true;
+}
```

Comment:
Maybe we can have a helper called `str::free()` that both drops the chars and frees the string itself. Is there a use case where we would want to drop but not free? If not, we can just make this one free the `str` object too by calling `free(self)`

Reply:
Okay